Sources/PackageSyntax/Formatting.swift

```diff
@@ -0,0 +1,199 @@
+        let formattedLastElement = formattingVisitor.visit(lastElement).as(ArrayElementSyntax.self)!
+        return self.withElements(elements.replacing(childAt: elements.count - 1, with: formattedLastElement))
+    }
+}
+
+fileprivate extension TriviaPiece {
+    var isComment: Bool {
+        switch self {
+        case .spaces, .tabs, .verticalTabs, .formfeeds, .newlines,
+             .carriageReturns, .carriageReturnLineFeeds, .garbageText:
+            return false
+        case .lineComment, .blockComment, .docLineComment, .docBlockComment:
+            return true
+        }
+    }
+
+    var isHorizontalWhitespace: Bool {
+        switch self {
+        case .spaces, .tabs:
+            return true
+        default:
+            return false
+        }
+    }
+
+    var isSpaces: Bool {
+        guard case .spaces = self else { return false }
+        return true
+    }
+
+    var isTabs: Bool {
+        guard case .tabs = self else { return false }
+        return true
+    }
+}
+
+fileprivate extension Trivia {
+    func droppingPiecesAfterLastComment() -> Trivia {
+        Trivia(pieces: .init(self.lazy.reversed().drop(while: { !$0.isComment }).reversed()))
+    }
+
+    func droppingPiecesUpToAndIncludingLastComment() -> Trivia {
+        Trivia(pieces: .init(self.lazy.reversed().prefix(while: { !$0.isComment }).reversed()))
+    }
+}
+
+extension SyntaxProtocol {
+    func determineIndentOfStartingLine() -> (indent: Trivia, unitIndent: Trivia) {
+        let sourceLocationConverter = SourceLocationConverter(file: "", tree: self.root.as(SourceFileSyntax.self)!)
+        let line = startLocation(converter: sourceLocationConverter).line ?? 0
+        let visitor = DetermineLineIndentVisitor(lineNumber: line, sourceLocationConverter: sourceLocationConverter)
+        visitor.walk(self.root)
+        return (indent: visitor.lineIndent, unitIndent: visitor.lineUnitIndent)
+    }
+}
+
+public final class DetermineLineIndentVisitor: SyntaxVisitor {
+
+    let lineNumber: Int
+    let locationConverter: SourceLocationConverter
+    private var bestMatch: TokenSyntax?
+
+    public var lineIndent: Trivia {
+        guard let pieces = bestMatch?.leadingTrivia
+                .lazy
+                .reversed()
+                .prefix(while: \.isHorizontalWhitespace)
+                .reversed() else { return .spaces(4) }
+        return Trivia(pieces: Array(pieces))
+    }
+
+    public var lineUnitIndent: Trivia {
+        if lineIndent.allSatisfy(\.isSpaces) {
+            let addedSpaces = lineIndent.reduce(0, {
+                guard case .spaces(let count) = $1 else { fatalError() }
+                return $0 + count
+            }) % 4 == 0 ? 4 : 2
+            return .spaces(addedSpaces)
+        } else if lineIndent.allSatisfy(\.isTabs) {
+            return .tabs(1)
+        } else {
+            // If we can't determine the indent, default to 4 spaces.
+            return .spaces(4)
```

Comment:
Obviously a nit and not for this PR, but in a future refinement we'd probably want to parameterize this.  Or do the heuristics for determining the number of spaces work so well that it isn't likely to be needed?

Reply:
The heuristic is heavily biased towards 4-spaces over 2-spaces right now, so I should definitely make this configurable at some point. 4 spaces is the default at the moment because that's what swift package --init and Xcode use.